### PR TITLE
ManifestLevel: SegmentedArray binary search

### PR DESCRIPTION
**Pull request by sentientwaffle**
#### Overview

- Replace `ManifestLevel`'s existing "root key" index structure with a simple two-phase binary search over the `SegmentedArray`.
  - Phase one: binary search the first key in each node.
  - Phase two: binary search the selected node's elements.
- Use binary search to compute the cursor from an absolute index. (Previously was a linear scan).
- Remove the `SegmentedArray.iterator()` `start_node` optimization.

#### Motivcation
- Without the necessity of amortizing the cost of rebuilding `ManifestLevel`'s root key index, `Manifest`/`ManifestLevel` updates no longer need to be batched. This simplifies the `Manifest` and `Compaction`.
- The old index structure was quite complex, even rebuilding it from scratch rather than incrementally.

#### Performance

- For "small" manifest levels, e.g. ~4k tables, `SegmentedArray.search` is faster.
- The original `ManifestLevel` search is around 2x faster at "large" levels.
- If `ManifestLevel` ends up a bottleneck, it will be simpler to implement an index (or other optimizations) within `SegmentedArray`.

### CODE REVIEW

File: src/lsm/segmented_array.zig [hunk 1]
```diff
@@ -679,13 +680,63 @@ pub fn SegmentedArray(
...
+                };
+            }
+
+            var offset: usize = 0;
+            var length: usize = array.node_count;
+            while (length > 1) {
+                const half = length / 2;
+                const mid = offset + half;
+
+                const node = &array.nodes[mid].?[0];
```
Comment (ifreund) [inline src/lsm/segmented_array.zig]: I wonder if explicit prefetching would make a bigger difference here than it did in our `binary_search.zig` implementation on contiguous arrays due to the additional indirection.

File: src/lsm/manifest_level_benchmark.zig [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            manifest_level.insert_tables(&node_pool, &tables);
+        }
+
+        const queries = try alloc_shuffled_index(allocator, options.table_count, prng.random());
+        defer allocator.free(queries);
+
+        const timer = try std.time.Timer.start();
+
+        const repetitions = std.math.max(1, @divFloor(samples, queries.len));
+        // Sum the results, just to make sure the compiler doesn't optimize all of this out.
```
Comment (ifreund) [inline src/lsm/manifest_level_benchmark.zig]: We could replace this with `std.mem.doNotOptimizeAway(absolute_index)` in the loop.

File: src/lsm/manifest_level.zig [hunk 3]
```diff
@@ -23,19 +24,20 @@ pub fn ManifestLevelType(
...
+                }
+            }.key_from_value,
+            compare_keys,
+        );
 
-        /// This is the index of the first table node that might contain the TableInfo
-        /// corresponding to a given key node. This allows us to skip table nodes which cannot
-        /// contain the target TableInfo when searching for the TableInfo with a given absolute
-        /// index. Only the first keys.node_count elements are valid.
-        root_table_nodes_array: *[Keys.node_count_max]u32,
```
Comment (ifreund) [inline src/lsm/manifest_level.zig:38]: While we can re-implement the `root_keys_array` "index" optimization to accelerate searching the segmented array in SegmentedArray itself or a thin wrapper, this optimization to convert from key node to table node in O(1) requires knowledge of both the keys and tables segmented arrays.

I'm not sure how much this matters in practice though due to your introduction of binary search in `SegmentedArray.cursor_for_absolute_index`. This is another place where I think we might have worse performance with very large levels and It doesn't seem to be compared in the current benchmarks.

**Review by ifreund — APPROVED** [review]
My overall impression is that this is a massive improvement in code maintainability and that we should merge this despite the performance deficit.

If the ManifestLevel turns out to be a bottleneck I'm also not too concerned about our ability to re-implement some or all of these optimizations, and If that turns out to be the case I believe we can do so without bringing back as much complexity as these commits remove.

Nice work!

File: src/lsm/segmented_array.zig (hunk 1 above)
Comment (sentientwaffle) [reply src/lsm/segmented_array.zig]: Good idea; I'll give it a try!

Comment (sentientwaffle) [reply src/lsm/segmented_array.zig]: Alas, no difference. I tried:


```
@prefetch(&array.nodes[offset + length / 2], .{});
```

and

```
@prefetch(&array.nodes[offset + length / 2], .{ .locality = 0 });
```

(at the end of the loop).

**Pull request merged**